src/dataflow/src/logging/rdkafka.rs

```diff
@@ -0,0 +1,177 @@
+}
+
+impl Peek {
+    /// Create a new peek from its arguments.
+    pub fn new(id: GlobalId, time: Timestamp, conn_id: u32) -> Self {
+        Self { id, time, conn_id }
+    }
+}
+
+/// Constructs the logging dataflows and returns a logger and trace handles.
+pub fn construct<A: Allocate>(
+    worker: &mut timely::worker::Worker<A>,
+    config: &dataflow_types::logging::LoggingConfig,
+    linked: std::rc::Rc<EventLink<Timestamp, (Duration, WorkerIdentifier, RDKafkaEvent)>>,
+) -> std::collections::HashMap<LogVariant, (Vec<usize>, KeysValsHandle)> {
+    let granularity_ms = std::cmp::max(1, config.granularity_ns / 1_000_000) as Timestamp;
+
+    let traces = worker.dataflow_named("Dataflow: rdkafka logging", move |scope| {
+        use differential_dataflow::collection::AsCollection;
+        use timely::dataflow::operators::capture::Replay;
+
+        let logs = Some(linked).replay_core(
+            scope,
+            Some(Duration::from_nanos(config.granularity_ns as u64)),
+        );
+
+        use timely::dataflow::operators::generic::builder_rc::OperatorBuilder;
+
+        let mut demux = OperatorBuilder::new("RDKafka Logging Demux".to_string(), scope.clone());
+        use timely::dataflow::channels::pact::Pipeline;
+        let mut input = demux.new_input(&logs, Pipeline);
+        let (mut consumer_info_out, consumer_info) = demux.new_output();
+
+        let mut demux_buffer = Vec::new();
+        demux.build(move |_capability| {
+            move |_frontiers| {
+                let mut consumer_info = consumer_info_out.activate();
+
+                input.for_each(|time, data| {
+                    data.swap(&mut demux_buffer);
+
+                    let mut consumer_info_session = consumer_info.session(&time);
+
+                    for (time, _, datum) in demux_buffer.drain(..) {
+                        let time_ns = time.as_nanos() as Timestamp;
+                        let time_ms = (time_ns / 1_000_000) as Timestamp;
+                        let time_ms = ((time_ms / granularity_ms) + 1) * granularity_ms;
+                        let time_ms = time_ms as Timestamp;
+
+                        match datum {
+                            RDKafkaEvent::ConsumerInfo {
+                                source_name,
+                                source_id,
+                                consumer_name,
+                                rx,
+                                rx_bytes,
+                                tx,
+                                tx_bytes,
+                            } => consumer_info_session.give((
+                                (
+                                    source_name,
+                                    source_id,
+                                    consumer_name,
+                                    rx,
+                                    rx_bytes,
+                                    tx,
+                                    tx_bytes,
+                                ),
```

Comment:
This seems fine, but as you're not using the tuples below (for, e.g. joins with other dataflows), you can save yourself one repetition of all those names below if you `.give` the packed row directly here (see timely.rs for an example of things doing that).

Reply:
Thanks! I'll look into that!